### PR TITLE
Check Ping return code and add fallback path for ping errors

### PR DESCRIPTION
**Pull request by strseb**
RE: https://github.com/mozilla-mobile/mozilla-vpn-client/issues/1495

From the new logs of luminita, the ping process exits with code 2 - (an error). We should check and handle it. 
My idea here is whenever we have such a ping error once a session, let's switch to the DUMMY_PING sender, so we don't do unnecessary steps here, like opening sockets and starting processes :) .

### CODE REVIEW

File: src/pingsender.cpp [hunk 1]
```diff
@@ -61,13 +61,22 @@ void PingSender::genericSendPing(const QStringList& args, qint16 sequence) {
 
   connect(process, qOverload<int, QProcess::ExitStatus>(&QProcess::finished),
           this, [this, sequence](int exitCode, QProcess::ExitStatus) {
-            if (exitCode == 0) emit recvPing(sequence);
+            logger.info() << "ping exit code: " << exitCode;
```
Comment (bakulf) [inline src/pingsender.cpp:64]: This is going to be super spammy. Remove this line

File: src/pinghelper.cpp [hunk 2]
```diff
@@ -50,17 +51,25 @@ void PingHelper::start(const QString& serverIpv4Gateway,
 
   m_gateway = serverIpv4Gateway;
   m_source = deviceIpv4Address.section('/', 0, 0);
-  m_pingSender =
+
+  if (s_has_critical_ping_error) {
+    logger.info() << "Encountered Unrecoverable ping error, using DUMMY Ping";
+    m_pingSender = new DummyPingSender(m_source, this);
```
Comment (bakulf) [inline src/pinghelper.cpp:57]: I like the solution, but we should go back to the normal ping sender after a while. Maybe after a few minutes?

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/pinghelper.cpp (hunk 2 above)
Comment (strseb) [reply src/pinghelper.cpp:58]: Good idea! Done :)

File: src/pinghelper.cpp [hunk 3]
```diff
@@ -192,3 +202,23 @@ double PingHelper::loss() const {
...
+  // and restart the pinghelper, to replace the impl with a dummy impl
+  // which we will use for the rest of the session
+  emit pingSentAndReceived(1);  // Fake a ping response with 1ms;
+  s_has_critical_ping_error = true;
+  stop();
+  start(m_gateway, m_source);
+
+  TimerSingleShot::create(qApp, 600000, [&] {  // 10 Minutes
+    logger.debug() << "Removing ping error state";
+    s_has_critical_ping_error = false;
```
Comment (bakulf) [inline src/pinghelper.cpp:222]: Here you have to call stop() + star(), right? Plus, if you set, as `parent` the m_pingSender, you don't need to check if the pingSender has been deleted in the meantime.

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/pinghelper.cpp [hunk 4]
```diff
@@ -192,3 +201,25 @@ double PingHelper::loss() const {
...
+  emit pingSentAndReceived(1);  // Fake a ping response with 1ms;
+  s_has_critical_ping_error = true;
+  stop();
+  start(m_gateway, m_source);
+
+  TimerSingleShot::create(this, 600000, [&] {  // 10 Minutes
+    logger.debug() << "Removing ping error state";
+    s_has_critical_ping_error = false;
+    stop();
+    start(m_gateway, m_source);
```
Comment (bakulf) [inline src/pinghelper.cpp:223]: What about if the time has been stopped in the meantime?

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/pinghelper.cpp (hunk 4 above)
Comment (strseb) [reply src/pinghelper.cpp]: good point lets just not do that :)

**Review by bakulf — APPROVED** [review]

**Pull request merged**